### DISCUSSION TITLE
How to ensure webdriverio is really getting page's elements

### DISCUSSION THREAD

**Discussion by kaabsimas** [webdriverio/webdriverio#8187] [General]
I've been facing an issue testing an Vue/Nuxt app with WebdriverIO. Mainly, it my tests are always passing. The following code will have an success execution:
```
it('should open correct app', () => {
    Page.open()
    browser.pause(1000)
    expect(browser).toHaveTitle('Checkout')
    browser.pause(1000)
  })
``` 

As does this one:
```
it('should open correct app', () => {
    Page.open()
    browser.pause(1000)
    expect(browser).toHaveTitle('Something Completely Different')
    browser.pause(1000)
  })
``` 

My first guess is that Page.open() is loading asynchronous, and the expect condition were being verified before the element existed. But if I try make the function asynchronous, to use await:
```
it('should open correct app', async () => {
    await Page.open()
    await browser.pause(1000)
    await expect(browser).toHaveTitle('Checkout')
    await browser.pause(1000)
  })
``` 
it gives me an error... of connection.

```
[Chrome 100.0.4896.60 linux #0-0] 1) Yourapp should open correct app
[Chrome 100.0.4896.60 linux #0-0] net::ERR_CONNECTION_REFUSED at http://localhost:3000/
[Chrome 100.0.4896.60 linux #0-0] net::ERR_CONNECTION_REFUSED at http://localhost:3000/
[Chrome 100.0.4896.60 linux #0-0] Error: net::ERR_CONNECTION_REFUSED at http://localhost:3000/
```
 
It happens because my tests are configured with an initialization of Nuxt server on the webdriver config hooks, that is basically inspired in [that](https://dev.to/astagi/testing-a-nuxt-js-application-using-webdriverio-1215) example:

wdio.conf.ts:
```
...
   async onPrepare () {
    if (process.env.NUXT_ENV_CI !== 'true') {
      return
    }
    const rootDir = path.resolve(__dirname, '.')
    const nuxtConfig = {
      head: {
        title: 'Yourapp'
      },
      dev: false,
      rootDir,
      modules: ['bootstrap-vue/nuxt']
    }
    this.nuxtInstance = new nuxt.Nuxt(nuxtConfig)
    await new nuxt.Builder(this.nuxtInstance).build()
    // its based on Express             ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
    await this.nuxtInstance.server.listen(3000, 'localhost')
  },
  onComplete () {
    if (process.env.NUXT_ENV_CI !== 'true') {
      return
    }
    this.nuxtInstance.close()
  }
...
```

and, despite it has been issued and corrected here https://github.com/webdriverio/webdriverio/issues/4042 looks like the asynchronous function is running after the onComplete hook is triggered. That's an issue I need the most to fix, since I'm using Browserstack Automate to test in multiple browsers in CI/CD and it only works when those hooks are correctly executed .

Anyway, I handled that locally running `npm run dev` on another terminal, what makes the tests run again, but they keep always passing successfully even when they shouldn't .

I'm using 

- Node 16.14.2
- NPM 8.5.0
- WebdriverIO 7.16.14
- Nuxt 2.15.8
- Ubuntu 20.04.4

**Comment by christian-bromann** [comment]
Can you provide a reproducible example? All hooks support async execution.

**Reply by kaabsimas** [reply]
Thanks Christian, I imported the test to a new project and pushed to a new repo to make an example. Its here:
https://github.com/kaabsimas/issue-example

**Reply by christian-bromann** [reply]
@kaabsimas you are not using `async` / `await` in your tests which is why all commands are run at the same time without waiting for their result.

**Reply by kaabsimas** [reply]
I mentioned that I tried using async/await. It does not solve the problem

**Comment by christian-bromann** [comment]
It seems that WebdriverIO has suppressed an error message:
```
Error in hook: Error: No `pages` directory found in /private/tmp/issue-example/test. Did you mean to run `nuxt` in the parent (`../`) directory?
    at Builder.validatePages (/private/tmp/issue-example/node_modules/@nuxt/builder/dist/builder.js:351:13)
    at async Builder.build (/private/tmp/issue-example/node_modules/@nuxt/builder/dist/builder.js:298:5)
    at async Object.onPrepare (/private/tmp/issue-example/test/wdio.conf.ts:165:5)
    at async Promise.all (index 0)
    at async Launcher.run (/private/tmp/issue-example/node_modules/@wdio/cli/build/launcher.js:85:13)
```

I created an issue in #8191. Any contributions are appreciated.
